xbmc/utils/ColorUtils.cpp

```diff
@@ -9,34 +9,165 @@
+  double b = (colorEx.colorARGB & 0x000000FF);
+  r /= 255;
+  g /= 255;
+  b /= 255;
+  const double& maxVal = std::max<double>({r, g, b});
+  const double& minVal = std::min<double>({r, g, b});
+  double h = 0;
+  double s = 0;
+  double l = (minVal + maxVal) / 2;
+  double d = maxVal - minVal;
+
+  if (d == 0)
+  {
+    h = s = 0; // achromatic
+  }
+  else
+  {
+    s = l > 0.5 ? d / (2 - maxVal - minVal) : d / (maxVal + minVal);
+    if (maxVal == r)
+    {
+      h = (g - b) / d + (g < b ? 6 : 0);
+    }
+    else if (maxVal == g)
+    {
+      h = (b - r) / d + 2;
+    }
+    else if (maxVal == b)
+    {
+      h = (r - g) / d + 4;
+    }
+    h /= 6;
+  }
+
+  colorEx.hue = h;
+  colorEx.saturation = s;
+  colorEx.lightness = l;
+}
+
+} // namespace
+
+Color COLORUTILS::ChangeOpacity(const Color color, const float opacity)
+{
+  int newAlpha = ceil(((color >> 24) & 0xff) * opacity);
   return (color & 0x00FFFFFF) | (newAlpha << 24);
 };
 
-UTILS::Color ColorUtils::ConvertToRGBA(const UTILS::Color argb)
+Color COLORUTILS::ConvertToRGBA(const Color argb)
 {
   return ((argb & 0x00FF0000) << 8) | //RR______
          ((argb & 0x0000FF00) << 8) | //__GG____
          ((argb & 0x000000FF) << 8) | //____BB__
          ((argb & 0xFF000000) >> 24); //______AA
 }
 
-UTILS::Color ColorUtils::ConvertToARGB(const UTILS::Color rgba)
+Color COLORUTILS::ConvertToARGB(const Color rgba)
 {
   return ((rgba & 0x000000FF) << 24) | //AA_____
          ((rgba & 0xFF000000) >> 8) | //__RR____
          ((rgba & 0x00FF0000) >> 8) | //____GG__
          ((rgba & 0x0000FF00) >> 8); //______BB
 }
 
-UTILS::Color ColorUtils::ConvertToBGR(const UTILS::Color argb)
+Color COLORUTILS::ConvertToBGR(const Color argb)
 {
   return (argb & 0x00FF0000) >> 16 | //____RR
          (argb & 0x0000FF00) | //__GG__
          (argb & 0x000000FF) << 16; //BB____
 }
+
+Color COLORUTILS::ConvertHexToColor(const std::string& hexColor)
+{
+  Color value = 0;
+  sscanf(hexColor.c_str(), "%x", &value);
+  return value;
+}
+
+ColorEx COLORUTILS::MakeColorEx(const Color& argb)
+{
+  ColorEx colorEx;
+  colorEx.colorARGB = argb;
+  colorEx.hex = StringUtils::Format("{:8x}", argb);
+  GetHSLValues(colorEx);
+  return colorEx;
+}
+
+ColorEx COLORUTILS::MakeColorEx(const std::string& hexColor)
+{
+  ColorEx colorEx;
+  colorEx.colorARGB = COLORUTILS::ConvertHexToColor(hexColor);
+  colorEx.hex = hexColor;
+  GetHSLValues(colorEx);
+  return colorEx;
+}
+
+bool COLORUTILS::comparePairColorEx(const std::pair<std::string, ColorEx>& a,
+                                    const std::pair<std::string, ColorEx>& b)
+{
+  if (a.second.hue == b.second.hue)
+  {
+    if (a.second.saturation == b.second.saturation)
+      return (a.second.lightness < b.second.lightness);
+    else
+      return (a.second.saturation < b.second.saturation);
+  }
+  else
+    return (a.second.hue < b.second.hue);
+}
+
+bool COLORUTILS::LoadXMLFileColors(const std::string& filePath,
```

Comment:
Loading colours from an XML file isn't really something I would think as an utils method. This likely belongs elsewhere although I can't say for sure where. We have a color manager but effort has been made into getting rid of it and moving it to the skin. Subtitle colours however belong to the system scope and should be always available (i.e. they don't depend depend on GUI or skins to be present (in an headless scenario we still want users to be able to configure the subtitles color as it might be present on external players). I'd try to move this into colormanager for now, or create a ColorLoader class and invoke it wherever this is actually needed

Reply:
i have moved it in to colormanager try give a look